### ISSUE TITLE
Minor float imprecision with String.split_floats()?

### ISSUE THREAD

**Issue by vybr** [godotengine/godot#14910]
**Godot version:**
Windows 3.0 Beta 1

**Issue description:**
```
var a = "4, 89.4, 5"
print(a.split_floats(","))
```
This code for some reason outputs [4, 89.400002, 5] instead of just [4, 89.4, 5]
When 89.42 is used it outputs 89.419998.

**Comment by poke1024** [comment]
A general problem with the accuracy of 32-bit floats as far as I see.

Note that in C, `printf("%f\n", 89.4f);` outputs `89.400002`.

In gdscript, the number gets gets stored in a `PoolRealArray` as a float (even though it's parsed inside `built_in_strtod` as a double). Once it's a float, that's it.

Another way to see that this happens during float conversion is adding

```
printf("v1 %lf\n", fraction);
float tmp1 = fraction;
printf("v2 %f\n", tmp1);
```

to the bottom of `built_in_strtod`. You get:

```
v1 89.400000
v2 89.400002
```

**Comment by Sslaxx** [comment]
Still reproducible with 3.2.2, but as mentioned above this is probably not easily fixable due to precision issues with floating-point type variables.

Unlikely I know, but would 4.0's double-precision compiling option "fix" this? And/or should this be closed as not really a Godot bug per se?

**Comment by Calinou** [comment]
I can still reproduce this on Godot 3.3.3: 
![image](https://user-images.githubusercontent.com/180032/132076166-ea4f199e-1f0f-4b45-a4bd-8866df9606c5.png)

**Comment by Mickeon** [comment]
Still happens as of b9a64c77366361a0d45dcdd2b330911efd1eb3f4, but the number is closer to **89.4**.
![image](https://user-images.githubusercontent.com/66727710/192307643-2f90217a-4b5b-4ac9-96ba-ee92c10b437a.png)

(Compiled with `scons -j6 platform=windows`)

**Comment by Mickeon** [comment]
The main reason of this issue is because `String.split_floats()` returns **PackedFloat32Array**, but the **float** type in GDScript is 64 bits. The method is used quite frequently internally, so I am not sure it is a good idea to double the precision, but that would actually solve the issue.

**Comment by Calinou** [comment]
> The main reason of this issue is because `String.split_floats()` returns **PackedFloat32Array**, but the **float** type in GDScript is 64 bits. The method is used quite frequently internally, so I am not sure it is a good idea to double the precision, but that would actually solve the issue.

Could we do this only when the engine is built with `float=64`?

**Comment by Mickeon** [comment]
The method could also be exposed **PackedFloat64Array**, but honestly, as I brought up, I do not believe this method to be particularly useful and it may be worth to just keep it unexposed, but...

Upon a second look, the method isn't used internally nearly as much as first assumed...
https://github.com/godotengine/godot/blob/8aafcf9d2ae43e83636d60dbfd5f10e50fcdb324/editor/plugins/script_text_editor.cpp#L1860-L1868
https://github.com/godotengine/godot/blob/8aafcf9d2ae43e83636d60dbfd5f10e50fcdb324/editor/editor_help.cpp#L1364-L1371

Other lines use a variation of the method prefixed with `_mk`.

So doubling the precision wouldn't be nearly as bad as first assumed...

**Comment by akien-mga** [comment]
> > The main reason of this issue is because `String.split_floats()` returns **PackedFloat32Array**, but the **float** type in GDScript is 64 bits. The method is used quite frequently internally, so I am not sure it is a good idea to double the precision, but that would actually solve the issue.
> 
> Could we do this only when the engine is built with `float=64`?

No, that's furthering a misunderstanding. `float=64` affects the precision of `Vector2/3/4` and other math types. It does not affect the precision of `float` which is always a C++ `double`.

So `String.split_floats()` should always return either an array of `float`s (i.e. 64-bit), or a `PackedFloat64Array`.

**Comment by akien-mga** [comment]
Fixed by #68917.
